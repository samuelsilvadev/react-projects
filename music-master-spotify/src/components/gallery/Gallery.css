.gallery {
    margin-top: 5rem;
}

.gallery__album {
    border: .1rem solid var(--white);
    cursor: pointer;
    display: inline-block;
    max-height: 220px;
    margin: .5rem;
    position: relative;
    text-align: center;
    width: 220px;
}

.gallery__album__image {
    object-fit: cover; 
    max-width: 100%; 
    width: 100%;
    height: inherit;
}

.gallery__album__title {
    background-color: var(--black);
    bottom: 0;
    left: 0;
    position: absolute;
    padding: 1rem;
    opacity: 0.75;
    width: 100%;
}