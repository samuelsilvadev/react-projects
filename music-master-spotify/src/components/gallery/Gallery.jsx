import React from 'react';

import './Gallery.css';
import { PropTypes } from 'prop-types';

class Gallery extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			audio: null,
			isPlayingAudio: false,
			previewUrl: ''
		};
	}

	componentWillUnmount() {
		if (this.state.audio) {
			this.state.audio.pause();
		}
	}

	render() {
		const { tracks } = this.props;

		return (
			<section className="gallery">
				<h3>Gallery</h3>
				{this._renderAlbums(tracks)}
			</section>
		);
	}

	_renderAlbums(tracks) {
		return (
			tracks &&
			tracks.map((track, index) => {
				const { album } = track;
				const [ imageAlbum ] = album.images;
				return (
					<div className="gallery__album" key={index} onClick={this._playAudio.bind(this, track.preview_url)}>
						<img
							className="gallery__album__image"
							src={imageAlbum.url}
							alt={album.name}
							title={album.name}
						/>
						<span className="gallery__album__title">{album.name}</span>
					</div>
				);
			})
		);
	}

	_playAudio(url) {
		const audio = new Audio(url);
		if (!this.state.isPlayingAudio) {
			audio.play();
			this.setState({
				isPlayingAudio: true,
				previewUrl: url,
				audio
			});
		} else {
			this.state.audio.pause();
			if (this.state.previewUrl === url) {
				this.setState({
					isPlayingAudio: false,
					previewUrl: '',
					audio: null
				});
			} else {
				audio.play();
				this.setState({
					isPlayingAudio: true,
					previewUrl: url,
					audio
				});
			}
		}
	}
}

Gallery.propTypes = {
	tracks: PropTypes.array
};

Gallery.defaultProps = {
	tracks: []
};

export default Gallery;
