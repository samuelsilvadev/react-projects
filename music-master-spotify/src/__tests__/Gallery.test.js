import React from 'react';
import { shallow } from 'enzyme';

import Gallery from '../components/gallery/Gallery';

describe('<Gallery />', () => {
	it('should render correctly component', () => {
		const tree = shallow(<Gallery />);

		expect(tree).toMatchSnapshot();
	});
});
