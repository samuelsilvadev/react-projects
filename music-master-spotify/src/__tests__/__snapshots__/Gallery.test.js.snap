// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Gallery /> should render correctly component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Gallery
    tracks={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h3>
          Gallery
        </h3>,
        Array [],
      ],
      "className": "gallery",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Gallery",
        },
        "ref": null,
        "rendered": "Gallery",
        "type": "h3",
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3>
            Gallery
          </h3>,
          Array [],
        ],
        "className": "gallery",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Gallery",
          },
          "ref": null,
          "rendered": "Gallery",
          "type": "h3",
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
